blocks/generator.go

```diff
@@ -41,6 +55,13 @@ func defaultConfig() Config {
 // GeneratorOpt for configuring Generator.
 type GeneratorOpt func(h *Generator)
 
+// WithContext modifies parent context that is used for all operations in Sync.
+func WithContext(ctx context.Context) GeneratorOpt {
```

Comment:
I don't really understand how this will be used. And should we really modify a context?
I feel like this should only ever be called once. But I do like the something.WithContext semantics.
Would a separate options struct be better than an abstract GeneratorOpt function?
I don't like the idea of Generator being mutable.
All these WithSomething functions should go on a config that can then be passed into the Generator via constructor.

Reply:
you can see the usage in cmd/node/node.go

this is called golang functional options pattern. it's not about the Generator being mutable or not. 
it's about how to configure Generator in a readable but non-repetitive way. 

config is an option, but you would end up with very long struct init code. 
with this pattern, the struct just need to init the most basic default stuff, so tests do not need to pass in any options.
only when you actually need to configure the binary do you need to call the WithXXX().

this pattern is generally considered superior as i know it. and our current codebase uses it extensively.